### ISSUE TITLE
Allow for token providers other than IAM

### ISSUE THREAD

**Issue by dchristidis** [rucio/rucio#6630]
### Description

The current token procurement workflow is IAM-specific (e.g. specifying the audience during the client-credentials flow).

https://github.com/rucio/rucio/blob/a7f97edbddf96c0abb1739bacf35581838c73de8/lib/rucio/core/oidc.py#L137-L141

### Motivation

This is a prerequisite for adding support for CILogon and Keycloak.

### Change

Create some abstraction for `request_token()` to be able to call implementation-specific functions.

**Comment by vokac** [comment]
IAM-specific behavior should be fixed https://github.com/indigo-iam/iam/issues/429

**Comment by dougbenjamin** [comment]
when is it expected that this will be in next release?

**Comment by dchristidis** [comment]
Following the discussion today at the Rucio Community Workshop, it should be noted that this issue is about supporting the existence of other token providers in Rucio, but not their implementations themselves.

Thus, if there’s a community that would like to use CILogon or Keycloak or anything else, they should begin by implementing a function similar to `request_token()` that takes arbitrary `audience` and `scope` arguments and returns an appropriate token. This issue will then provide the ‘glue’ so that this new function can be used.

**Comment by tbeerman** [comment]
I would just like to add here that we are also waiting for this. We have instances both with EGI Checkin (Keycloak) and Helmholtz ID ([Unity IDM](https://unity-idm.eu/)) and I'm more than happy to write a `request_token` implementation for that. In our currently deployments I manually patched the code to get it working. But I don't think adapting `request_token` alone would be enough. AFAIK it is not possible to get arbitrary scopes, as you do in `determine_scope_for_rse` from Keycloak and Unity IDM. They have to be defined in the client beforehand. So I guess this also has to be adapted to get it working.

**Comment by jamesp-epcc** [comment]
I agree with Thomas that adapting `request_token` will not be enough on its own. In our tests for DUNE using CILogon, the code within `request_token` works fine as it is (at least on the test instance of CILogon), but we had to modify `__get_init_oidc_client` to remove the audience parameter before user authentication would work.

**Comment by vokac** [comment]
CILogon doesn't support "OAuth Resource Indicators" [RFC8707](https://datatracker.ietf.org/doc/html/rfc8707)? During last [IAM hackaton](https://indico.cern.ch/event/1443968/overview) I asked developer about https://github.com/indigo-iam/iam/issues/841, if I remember correctly it should not be difficult to support RFC way to get audience, but no one was assigned to this ticket.
